posts/2021-07-21-building-on-atomic-reference.adoc

```diff
@@ -0,0 +1,208 @@
+[source, java]
+----
+    @FunctionalInterface
+    private interface Ref<T> extends Supplier<T> {
+        default boolean isReady() { return true; }
+    }
+----
+.It was Greek to me
+****
+Marking the interface with `@FunctionalInterface` ensures it is lambda-capable.
+The compiler checks that the interface satisfies Java's lambda constraints.
+****
+While the update is being applied, we want a special `Ref` that makes other threads wait for the result:
+[source, java]
+----
+    /** A placeholder ref to allow contending threads to wait for an in-flight update. */
+    private static class TransitionalRef<T> implements Ref<T> {
+        private final CountDownLatch latch = new CountDownLatch(1);
+        private final Thread updaterThread = Thread.currentThread();
+        public boolean isReady() {
+            // It is not valid to try to retrieve the value
+            // from the updating thread while it is still updating
+            if (Thread.currentThread() == updaterThread)
+                throw new IllegalStateException("Attempt to retrieve value during update process");
+            try {
+                latch.await();
+            } catch (InterruptedException ignored) {}
+            return false;
+        }
+        void markAsReady() { latch.countDown(); }
+        public T get() { throw new UnsupportedOperationException();  }
+    }
+----
+We hold one of these `Ref` objects inside an `AtomicReference`.
+Instead of `AtomicReference<Widget>` we now use `AtomicReference<Ref<Widget>>`.
+
+=== And waits upon the judgement
+Waiting for a transition is very simple:
+[source, java]
+----
+    private Ref<T> getWithWait() {
+        for (;;) {
+            Ref<T> ref = atomicReference.get();
+            if (ref.isReady()) return ref;
+        }
+    }
+----
+Usually, this loop will run only once, because `ref.isReady()` returns true.
+If `ref` is a `TransitionalRef`, the `isReady()` method will block until the transition completes.
+Even then, the loop would usually repeat once more and then retrieve the updated value.
+An unlucky thread could loop around many times -- once for each of many observed transitions.
+
+.For ever and a day
+****
+Some programmers prefer the `while (true)` or `do`...`while (true)` style of indefinite loop.
+The empty `for` has fewer compare operations,
+and the `(;;)` syntax is more fun.
+We call it the spider operator. &#128375;
+****
+
+The `get()` method is trivial -- get the reference, and dereference it:
+[source, java]
+----
+    public T get() { return getWithWait().get(); }
+----
+
+=== What judgement would step from this to this?
+
+The `update()` implementation is more involved.
+Only one thread may enter the critical section where the update happens.
+Other threads must wait for the transition to complete before they can access the result.
+[source, java]
+----
+    public boolean update(Function<T, Supplier<T>> updaterSupplier) {
```

Comment:
The code callouts look fine. We avoid them in the docs because they are not accessible to screen readers. I'm not sure what the expectations/obligations w/r/t to accessibility are for the blog posts. 

@steven1046 do you know if callouts are ok in the blog posts? I can't remember the exact instance where they were deemed inaccessible in the docs.

Reply:
Yeah, i think callouts were not accessible. See https://github.com/OpenLiberty/docs/issues/3304 for having to remove them from docs.

Reply:
Do your recent CSS changes that went into staging render the callouts accessible, @steven1046?
I'm keen to preserve accessibility so I will re-write if necessary.

Reply:
The CSS changes were only for sidebars, and not callouts

Reply:
OK, if I understand correctly, it is just the code callouts that are inaccessible.
We will re-work these as code comments or interspersed descriptive text.